velox/exec/HashProbe.cpp

```diff
@@ -292,6 +292,53 @@ void HashProbe::maybeSetupSpillInputReader(
   inputSpillPartitionSet_.erase(iter);
 }
 
+std::optional<uint64_t> HashProbe::estimatedRowSize(
+    const std::vector<vector_size_t>& varSizedColumns,
+    uint64_t totalFixedColumnsBytes,
+    double toleranceRatio) {
+  std::vector<RowColumnStats> varSizeListColumnsStats;
+  varSizeListColumnsStats.reserve(varSizedColumns.size());
+  for (uint32_t i = 0; i < varSizedColumns.size(); i++) {
+    auto statsOpt = columnStats(varSizedColumns[i]);
+    if (!statsOpt.has_value()) {
+      return std::nullopt;
+    }
+    varSizeListColumnsStats.push_back(statsOpt.value());
+  }
+
+  uint64_t totalAvgBytes{totalFixedColumnsBytes};
+  uint64_t totalMaxBytes{totalFixedColumnsBytes};
+  for (const auto& stats : varSizeListColumnsStats) {
+    totalAvgBytes += stats.avgBytes();
+    totalMaxBytes += stats.maxBytes();
+  }
+  if (totalAvgBytes == 0) {
+    if (totalMaxBytes == 0) {
+      return 0;
+    }
+    return std::nullopt;
+  }
+  if (totalMaxBytes / totalAvgBytes < toleranceRatio) {
+    return (totalMaxBytes + totalAvgBytes) / 2;
+  }
+  return std::nullopt;
+}
+
+std::optional<RowColumnStats> HashProbe::columnStats(
+    int32_t columnIndex) const {
+  std::vector<RowColumnStats> columnStats;
```

Comment:
It is better to make this a method of RowContainer? Thanks!

Reply:
It consolidates multiple row containers stats so it might not be applicable to have it inside row container.